contracts/src/lordship.cairo

```diff
@@ -0,0 +1,245 @@
+        fn before_update(
+            ref self: ERC721Component::ComponentState<TContractState>,
+            to: ContractAddress,
+            token_id: u256,
+            auth: ContractAddress
+        ) {
+            let owner_before_transfer: ContractAddress = self.owner_of(token_id);
+            let owner_after_transfer: ContractAddress = to;
+
+            // claim stream for both sender and receiver
+            let mut strealm_component = get_dep_component_mut!(ref self, StRealm);
+            strealm_component._update_stream_balance(owner_before_transfer);
+            strealm_component._update_stream_balance(owner_after_transfer);
+
+            // transfer voting units 
+            let mut erc721_votes_component = get_dep_component_mut!(ref self, ERC721Votes);
+            erc721_votes_component.transfer_voting_units(owner_before_transfer, to, 1);
+        }
+
+        fn after_update(
+            ref self: ERC721Component::ComponentState<TContractState>,
+            to: ContractAddress,
+            token_id: u256,
+            auth: ContractAddress
+        ) {}
+    }
+
+
+    //
+    // Hooks ERC721VotesHooks
+    //
+
+    impl ERC721VotesHooksImpl<
+        TContractState,
+        impl StRealm: StRealmComponent::HasComponent<TContractState>,
+        impl HasComponent: ERC721VotesComponent::HasComponent<TContractState>,
+        +ERC721Component::HasComponent<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +NoncesComponent::HasComponent<TContractState>,
+        +Drop<TContractState>,
+    > of ERC721VotesComponent::ERC721VotesHooksTrait<TContractState> {
+        fn before_delegate(
+            ref self: ERC721VotesComponent::ComponentState<TContractState>,
+            account: ContractAddress,
+            delegatee: ContractAddress
+        ) {
+            let mut strealm_component = get_dep_component_mut!(ref self, StRealm);
+            if delegatee.is_zero() {
+                strealm_component._update_stream_balance(account);
+                strealm_component._end_stream(account);
+            } else {
+                if self.delegates(account).is_zero() {
+                    // no current delegates
+                    strealm_component._reset_stream(account);
+                }
+            }
+        }
+
+        fn after_delegate(
+            ref self: ERC721VotesComponent::ComponentState<TContractState>,
+            account: ContractAddress,
+            delegatee: ContractAddress
+        ) {}
+    }
+
+
+    #[abi(embed_v0)]
+    impl UpgradeableImpl of IUpgradeable<ContractState> {
+        fn upgrade(ref self: ContractState, new_class_hash: ClassHash) {
+            self.access_control.assert_only_role(UPGRADER_ROLE);
+            self.upgradeable._upgrade(new_class_hash);
+        }
+    }
+
+
+    #[generate_trait]
+    #[abi(per_item)]
+    impl ERC721MinterImpl of ERC721MinterTrait {
+        #[abi(embed_v0)]
```

Comment:
When using `abi(per_item)`, shouldn't this actually be `#[external(v0)]`?

Reply:
true